File: mobile/src/core/care/thermalShockGuard.ts
import { db } from "../storage/db";
import { protectionEvents, thermalEvents } from "../storage/schema";
import { boostKpa, type VehicleSnapshot } from "../obd/vehicleSnapshot";
import type { CareCue, CareContext } from "./careTypes";
import type { CareFeature } from "./careFeature";

const REDLINE_RPM = 6500;
const MAX_BOOST_KPA = 150;
/** Fixed profile factor (water-cooled turbo, no aux electric pump) — the unported
 * per-vehicle profile system would tune this per powertrain. */
const PROFILE_FACTOR = 0.5;

function computeTLI(samples: Array<{ t: number; v: number }>): number {
  if (samples.length < 2) return samples[samples.length - 1]?.v ?? 0;
  let sum = 0;
  for (let i = 1; i < samples.length; i++) {
    const dt = (samples[i]!.t - samples[i - 1]!.t) / 1000;
    sum += samples[i]!.v * Math.max(dt, 0);
  }
  return sum / 300;
}

function recommendedIdleS(tli: number): number {
  if (tli < 0.25) return 0;
  if (tli < 0.45) return 20;
  if (tli < 0.7) return 45;
  if (tli < 1.0) return 75;
  return 120;
}

/** Simplified port of ThermalShockGuard.swift — detects turbo/engine "hot shutdown"
 * risk and recommends an idle-before-shutoff cooldown window. */
class ThermalShockGuard implements CareFeature {
  id = "thermalShock";

  countdownSeconds: number | undefined;

  private heatSamples: Array<{ t: number; v: number }> = [];
  private stopSince: number | undefined;
  private countdownActive = false;
  private countdownRemaining = 0;
  private idleAccumulated = 0;
  private announcedStart = false;
  private announcedDone = false;

  isEnabled(settings: { careThermalShock: string }): boolean {
    return settings.careThermalShock !== "off";
  }

  evaluate(snapshot: VehicleSnapshot, context: CareContext): CareCue[] {
    const now = context.now;
    const rpm = snapshot.rpm ?? 0;
    const speed = snapshot.speedKmh ?? 0;
    const load = snapshot.engineLoadPct ?? 0;
    const boost = boostKpa(snapshot) ?? 0;

    const boostFactor = Math.min(Math.max(1.0 + Math.max(boost, 0) / MAX_BOOST_KPA, 1.0), 2.2);
    let instant = (rpm / REDLINE_RPM) * (load / 100) * boostFactor;
    if (context.oilTempC != null && context.oilTempC > 105) instant *= 1.15;
    this.heatSamples.push({ t: now, v: instant });
    this.heatSamples = this.heatSamples.filter((s) => now - s.t <= 300_000);

    const tli = computeTLI(this.heatSamples);
    const wait = recommendedIdleS(tli) * PROFILE_FACTOR;

    const cues: CareCue[] = [];

    if (speed < 2 && rpm > 300) {
      if (this.stopSince == null) this.stopSince = now;
      const stoppedForMs = now - this.stopSince;
      if (wait > 0 && stoppedForMs >= 20_000) {
        if (!this.countdownActive) {
          this.countdownActive = true;
          this.countdownRemaining = wait;
          this.idleAccumulated = 0;
          this.announcedStart = false;
          this.announcedDone = false;
        }
        this.idleAccumulated += 1;
        this.countdownRemaining = Math.max(0, wait - this.idleAccumulated);
        this.countdownSeconds = this.countdownRemaining;

        if (!this.announcedStart) {
          this.announcedStart = true;
          cues.push({ id: "thermal.start", text: `Let the engine idle ~${Math.round(wait)}s before shutting off to protect the turbo.`, severity: "protective" });
        }
        if (this.countdownRemaining <= 0 && !this.announcedDone) {
          this.announcedDone = true;
          cues.push({ id: "thermal.done", text: "Safe to turn off now.", severity: "celebration" });
          db.insert(thermalEvents)
            .values({ t: now, tli, recommendedIdleS: wait, actualIdleS: this.idleAccumulated, compliant: true })
            .catch(() => undefined);
        }
      }
    } else if (speed >= 2) {
      this.stopSince = undefined;
      if (this.countdownActive && this.countdownRemaining > 0) {
        this.countdownActive = false;
        this.countdownSeconds = undefined;
      }
    }

    if (rpm < 50 && this.countdownActive && this.countdownRemaining > 0) {
      db.insert(thermalEvents)
        .values({ t: now, tli, recommendedIdleS: wait, actualIdleS: this.idleAccumulated, compliant: false })
        .catch(() => undefined);
      db.insert(protectionEvents)
        .values({ id: `pe_${now}_${Math.random().toString(36).slice(2, 6)}`, type: "hotShutdown", severity: "alarm", t: now, value: tli, thresholdUsed: wait })
        .catch(() => undefined);
      this.countdownActive = false;
      this.countdownSeconds = undefined;
    }

    return cues;
  }

  resetTrip(): void {
    this.heatSamples = [];
    this.stopSince = undefined;
    this.countdownActive = false;
    this.countdownRemaining = 0;
    this.countdownSeconds = undefined;
  }
}

export const thermalShockGuard = new ThermalShockGuard();

File: QuickCar/Core/Care/ThermalShockGuard.swift
import Foundation
import SwiftData

@MainActor
final class ThermalShockGuard: CareFeature {
    let id = "thermalShock"
    let requiredPIDs: Set<UInt8> = [0x0C, 0x04]
    let optionalPIDs: Set<UInt8> = [0x0B, 0x33, 0x5C]

    private let modelContext: ModelContext
    private var heatSamples: [(Date, Double)] = []
    private var stopSince: Date?
    private var countdownActive = false
    private var countdownRemaining: Double = 0
    private var lastTLI: Double = 0
    private var recommendedIdle: Double = 0
    private var announcedStart = false
    private var announcedDone = false
    private var idleAccumulated: Double = 0
    private var regenHoldAnnounced = false

    var countdownSeconds: Double? { countdownActive ? countdownRemaining : nil }

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }

    func isEnabled(settings: AppSettings) -> Bool {
        settings.careThermalShock != "off"
    }

    func evaluate(snapshot: VehicleSnapshot, context: inout CareContext) -> [CareCue] {
        let profile = context.profile
        // Full hybrid: ICE durur/tekrar başlar, kontrollü kapanış zaten var — modül devre dışı.
        guard !profile.flags.contains("noThermalShockModule") else { return [] }

        let now = context.now
        let rpm = snapshot.rpm ?? 0
        let speed = snapshot.speedKmh ?? 0
        let load = snapshot.engineLoadPct ?? 0
        let boost = snapshot.boostKpa ?? 0

        let boostFactor: Double = profile.aspiration != .na
            ? clamp(1.0 + max(boost, 0) / max(profile.maxBoostKpa, 1), 1.0, 2.2)
            : 1.0
        var instant = (rpm / max(profile.redlineRpm, 1)) * (load / 100) * boostFactor
        if let oil = context.oilTempC, oil > 105 {
            instant *= 1.15
        }
        heatSamples.append((now, instant))
        heatSamples = heatSamples.filter { now.timeIntervalSince($0.0) <= 300 }

        let tli = Self.computeTLI(samples: heatSamples, now: now)
        lastTLI = tli
        var wait = Self.recommendedIdle(tli: tli) * Self.profileFactor(profile: profile)
        recommendedIdle = wait

        var cues: [CareCue] = []

        if profile.hasDPF, context.regenActive {
            wait = max(wait, 90)
            recommendedIdle = wait
            if !regenHoldAnnounced, speed < 2, rpm > 300 {
                regenHoldAnnounced = true
                cues.append(CareCue(
                    id: "thermal.regenHold",
                    text: LocalizedCare.string("thermal.countdown.voice", Int(wait.rounded())),
                    severity: .protective,
                    localizationKey: "thermal.countdown.voice"
                ))
            }
        } else {
            regenHoldAnnounced = false
        }

        if speed < 2, rpm > 300 {
            if stopSince == nil { stopSince = now }
            let stoppedFor = now.timeIntervalSince(stopSince ?? now)
            if wait > 0, stoppedFor >= 20 {
                if !countdownActive {
                    countdownActive = true
                    countdownRemaining = wait
                    idleAccumulated = 0
                    announcedStart = false
                    announcedDone = false
                }
                idleAccumulated += 1
                countdownRemaining = max(0, wait - idleAccumulated)
                context.thermalCountdownS = countdownRemaining

                if !announcedStart {
                    announcedStart = true
                    let secs = Int(wait.rounded())
                    cues.append(CareCue(
                        id: "thermal.start",
                        text: LocalizedCare.string("thermal.countdown.voice", secs),
                        severity: .protective,
                        localizationKey: "thermal.countdown.voice"
                    ))
                }
                if countdownRemaining <= 0, !announcedDone {
                    announcedDone = true
                    cues.append(CareCue(
                        id: "thermal.done",
                        text: LocalizedCare.string("thermal.done.voice"),
                        severity: .celebration,
                        localizationKey: "thermal.done.voice"
                    ))
                    modelContext.insert(ThermalEvent(
                        tli: tli, recommendedIdleS: wait, actualIdleS: idleAccumulated, compliant: true
                    ))
                    try? modelContext.save()
                }
            }
        } else if speed >= 2 {
            stopSince = nil
            if countdownActive, countdownRemaining > 0 {
                countdownActive = false
                context.thermalCountdownS = nil
            }
        }

        // Hot shutdown: rpm → 0 while cooldown incomplete.
        if rpm < 50, countdownActive, countdownRemaining > 0 {
            modelContext.insert(ThermalEvent(
                tli: tli, recommendedIdleS: wait, actualIdleS: idleAccumulated, compliant: false
            ))
            try? modelContext.save()
            if profile.hasAuxElectricCoolantPump.value {
                // Kontak sonrası elektrikli pompa işi yapar — bilgi amaçlı, ceza yok.
                cues.append(CareCue(
                    id: "thermal.hotShutdownSoft",
                    text: LocalizedCare.string("thermal.done.voice"),
                    severity: .coach,
                    localizationKey: "thermal.done.voice"
                ))
            } else {
                modelContext.insert(ProtectionEvent(
                    typeRaw: "hotShutdown", severityRaw: "alarm", value: tli, thresholdUsed: wait
                ))
                try? modelContext.save()
            }
            countdownActive = false
            context.thermalCountdownS = nil
        }

        return cues
    }

    /// §4 profileFactor — modern su soğutmalı turbo + kontak sonrası pompa büyük ölçüde
    /// önerilen bekleme süresini eritir; NA motorda öneri hiç üretilmez (sadece log).
    nonisolated static func profileFactor(profile: VehicleDiagnosticProfile) -> Double {
        if profile.fuel == .diesel, profile.hasDPF { return 0.6 }
        guard profile.aspiration != .na else { return 0.0 }
        if profile.turboWaterCooled, profile.hasAuxElectricCoolantPump.value { return 0.25 }
        if profile.turboWaterCooled { return 0.5 }
        return 1.0
    }

    nonisolated static func computeTLI(samples: [(Date, Double)], now: Date) -> Double {
        guard samples.count >= 2 else { return samples.last?.1 ?? 0 }
        var sum = 0.0
        for i in 1..<samples.count {
            let dt = samples[i].0.timeIntervalSince(samples[i - 1].0)
            sum += samples[i].1 * max(dt, 0)
        }
        return sum / 300
    }

    nonisolated static func recommendedIdle(tli: Double) -> Double {
        switch tli {
        case ..<0.25: return 0
        case 0.25..<0.45: return 20
        case 0.45..<0.70: return 45
        case 0.70..<1.00: return 75
        default: return 120
        }
    }

    private func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double {
        min(max(v, lo), hi)
    }
}
